### PR TITLE
feat: integrate the llama3 (8B, 70B), Mistral.AI, Gemma (7B, 9B) served by Groq

### PR DESCRIPTION
**Pull request by Appointat**
## Description

Describe your changes in detail.

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] New feature (non-breaking change which adds core functionality)
- [x] Example (update in the folder of example)

### CODE REVIEW

**Comment by Appointat** [comment]
@Wendong-Fan Hi, could you please help me add `GROQ_API_KEY` to the GitHub secret (I  will DM you the secrete key value)? Thanks a lot in advance.

File: camel/models/groq_model.py
Comment (ocss884) [inline camel/models/groq_model.py:62]: Why not use `Autotokenizer` to load the correct llama3 tokenizer?

**Review by ocss884 — COMMENTED** [review]
LGTM! Left a question

File: camel/models/groq_model.py
Comment (Appointat) [reply camel/models/groq_model.py:62]: Good idea, and I will try it.

Comment (Appointat) [reply camel/models/groq_model.py]: Hi, I have implemented the Autotokenizer. wait for your review

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @Appointat 's contribution! Overall is great, left some comments

File: camel/utils/token_counting.py
Comment (ocss884) [inline camel/utils/token_counting.py:316]: I suggest not launching any tokenizer if allowed. User won't go through every line of code but just find there is a tokenizer and use it. They can not notice it is not the correct one until they notice some errors in token number and try to figure out what's the problem.

Or, we need to specify which open-source model is launched and try to load it via HF libraries.

**Review by ocss884 — COMMENTED** [review]
left some comments

File: camel/models/groq_model.py
Comment (Appointat) [reply camel/models/groq_model.py:111]: Modified

**Comment by Appointat** [comment]
@camel-ai/camel-maintainers  @Wendong-Fan hi guys, this pr is fixed finally, could you please review it? thanks

File: camel/configs/__init__.py
Comment (Wendong-Fan) [inline camel/configs/__init__.py:36]: groq also support other model types beyond llama3, we can make this integration more general

File: camel/models/groq_model.py
Comment (Wendong-Fan) [inline camel/models/groq_model.py:89]: groq supports OpenAI compatibility, how about use openai interface rather than do the transform by ourselves?
https://console.groq.com/docs/openai

File: camel/configs/groq_llama3_config.py
Comment (Wendong-Fan) [inline camel/configs/groq_llama3_config.py:51]: we can use OpenAI compatibility, refer to https://console.groq.com/docs/openai

File: camel/models/model_factory.py
Comment (Wendong-Fan) [inline camel/models/model_factory.py:73]: set groq as platform, support other model types including mixtral-8x7b, refer to https://console.groq.com/docs/models

File: camel/utils/token_counting.py
Comment (Wendong-Fan) [inline camel/utils/token_counting.py:26]: set Groq and AutoTokenizer within the class to make it optional

File: camel/utils/token_counting.py
Comment (Wendong-Fan) [inline camel/utils/token_counting.py:335]: here I think it's OpenAI API format since it's OpenAIMessage

File: camel/utils/token_counting.py
Comment (Wendong-Fan) [inline camel/utils/token_counting.py:446]: if there's no token counter provided from Groq side, we can set the token counter under `OpenSourceTokenCounter`

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @Appointat 's update, left some comments

File: camel/configs/__init__.py
Comment (Appointat) [reply camel/configs/__init__.py:36]: Groq supports some famous open source models, but the configs of every model may not be the same. After we have integrated other models served by groq, we can improve this at that time. What do you think of it?

File: camel/models/groq_model.py
Comment (Appointat) [reply camel/models/groq_model.py:89]: I have checked the compatibility, but it can not pass the `mypy` check, according to the expression in groq: "However, there are some nuanced differences where support is not yet available."

File: camel/models/model_factory.py
Comment (Appointat) [reply camel/models/model_factory.py:73]: We can integrate the mistral later (next pr) otherwise, the workload will be doubled (since the support of groq is really bad)? What do you think of it?

File: camel/utils/token_counting.py
Comment (Appointat) [reply camel/utils/token_counting.py:26]: What do you mean optional? In my understanding, the optional is usually used for the parameters?

File: camel/configs/groq_llama3_config.py
Comment (Appointat) [reply camel/configs/groq_llama3_config.py:51]: The models served be groq do not support all the parameters of the specs of openai. @Wendong-Fan
https://console.groq.com/docs/api-reference#chat-create
1. tool call can be done in the next pr?
2. logit_bias, logprobs, top_logprobs not supported

File: camel/models/groq_model.py
Comment (Wendong-Fan) [reply camel/models/groq_model.py:89]: thanks for the reply, I think it's not related to `mypy` check, you can refer to `camel/configs/ollama_config.py` set a config for Groq with openai format

File: camel/configs/groq_llama3_config.py
Comment (Wendong-Fan) [reply camel/configs/groq_llama3_config.py:51]: we can just remove these parameters not supported, they are also not regularly used

File: camel/models/model_factory.py
Comment (Wendong-Fan) [reply camel/models/model_factory.py:73]: Hey @Appointat , I think the workload would not be double, just some small change, don't hard code the model name, refer to how Ollama was integrated : camel/models/ollama_model.py

File: camel/utils/token_counting.py
Comment (Wendong-Fan) [reply camel/utils/token_counting.py:26]: Screenshot 2024-07-14 at 20 01 32
I mean move the import within the class like here, so that the library will only be imported when the class is used

**Comment by Appointat** [comment]
@Wendong-Fan I have seen an error in pytest

```
FAILED test/models/test_groq_llama3_model.py::test_groq_llama3_model[ModelType.GROQ_LLAMA_3_8_B] - ValueError: Invalid `model_path` (meta-llama/Meta-Llama-3-8B-Instruct) is provided. Tokenizer loading failed.
FAILED test/models/test_groq_llama3_model.py::test_groq_llama3_model[ModelType.GROQ_LLAMA_3_70_B] - ValueError: Invalid `model_path` (meta-llama/Meta-Llama-3-70B-Instruct) is provided. Tokenizer loading failed.
```

It means the model can not download the tokenizer from huggingface, we need to add `HUGGING_FACE_HUB_TOKEN`. Could you please help me to apply one and add it to the env? thanks!

**Comment by Wendong-Fan** [comment]
token counter access issue fixed

**Comment by Appointat** [comment]
Thank you, I will fix the conflits.

**Comment by Wendong-Fan** [comment]
Hey @Appointat , did some update in https://github.com/camel-ai/camel/pull/531/commits/407b44e85fe1afb40ddcfb3575b19f198f6ea2f9

please review the change, I used OpenAITokenCounter as default token counter since it's more easy for user set it up even it's not accurate for open source models, now we also allow user to switch the token counter when initilizing the model, let me know WDYT~

**Review by Wendong-Fan — APPROVED** [review]

**Comment by Appointat** [comment]
@Wendong-Fan I have checked the code, but I cannot review the code. I think it is ok to be merged. Thank you.

**Pull request merged**
